File: server/src/lib/sources/hackernews.ts
import axios from 'axios'

export interface HNItem {
  title: string
  url: string
  content: string
  author: string
  publishedAt: Date
  source: 'hackernews'
  points: number
  numComments: number
}

const HN_API = 'https://hn.algolia.com/api/v1'

export async function searchHackerNews(query: string, hoursBack = 24): Promise<HNItem[]> {
  try {
    const sinceTs = Math.floor(Date.now() / 1000) - hoursBack * 3600
    const res = await axios.get(`${HN_API}/search_by_date`, {
      params: {
        query,
        tags: 'story',
        numericFilters: `created_at_i>${sinceTs}`,
        hitsPerPage: 20,
      },
      timeout: 10000,
    })

    return (res.data.hits || [])
      .filter((h: any) => h.title && (h.url || h.story_url))
      .map((h: any) => ({
        title: h.title,
        url: h.url || h.story_url || `https://news.ycombinator.com/item?id=${h.objectID}`,
        content: h.story_text || h.comment_text || h.title,
        author: h.author || '',
        publishedAt: new Date(h.created_at),
        source: 'hackernews' as const,
        points: h.points || 0,
        numComments: h.num_comments || 0,
        likes: h.points || 0,
        comments: h.num_comments || 0,
      }))
  } catch (err) {
    console.error('[HackerNews] search error:', err)
    return []
  }
}

export async function getHNTrending(query: string): Promise<HNItem[]> {
  try {
    const res = await axios.get(`${HN_API}/search`, {
      params: {
        query,
        tags: 'story',
        hitsPerPage: 30,
        numericFilters: `points>10,num_comments>5`,
      },
      timeout: 10000,
    })

    return (res.data.hits || [])
      .filter((h: any) => h.title && (h.url || h.story_url))
      .map((h: any) => ({
        title: h.title,
        url: h.url || h.story_url || `https://news.ycombinator.com/item?id=${h.objectID}`,
        content: h.story_text || h.title,
        author: h.author || '',
        publishedAt: new Date(h.created_at),
        source: 'hackernews' as const,
        points: h.points || 0,
        numComments: h.num_comments || 0,
        likes: h.points || 0,
        comments: h.num_comments || 0,
      }))
  } catch (err) {
    console.error('[HackerNews] trending error:', err)
    return []
  }
}

File: .cursor/skills/hot-monitor/scripts/lib/hackernews.py
from __future__ import annotations

from datetime import datetime, timezone

import requests

HN_API = "https://hn.algolia.com/api/v1"
UA = {"User-Agent": "HotMonitor-Skill/1.0"}


def _map_hit(h: dict) -> dict:
    url = h.get("url") or h.get("story_url") or f"https://news.ycombinator.com/item?id={h.get('objectID')}"
    created = h.get("created_at")
    pub = datetime.fromisoformat(created.replace("Z", "+00:00")) if created else datetime.now(timezone.utc)
    pts = h.get("points") or 0
    comments = h.get("num_comments") or 0
    return {
        "title": h.get("title", ""),
        "url": url,
        "content": h.get("story_text") or h.get("comment_text") or h.get("title", ""),
        "author": h.get("author") or "",
        "publishedAt": pub,
        "source": "hackernews",
        "likes": pts,
        "comments": comments,
    }


def search(query: str, hours_back: int = 24) -> list[dict]:
    since_ts = int(datetime.now(timezone.utc).timestamp()) - hours_back * 3600
    try:
        r = requests.get(
            f"{HN_API}/search_by_date",
            params={
                "query": query,
                "tags": "story",
                "numericFilters": f"created_at_i>{since_ts}",
                "hitsPerPage": 20,
            },
            headers=UA,
            timeout=10,
        )
        r.raise_for_status()
        hits = r.json().get("hits") or []
        return [
            _map_hit(h)
            for h in hits
            if h.get("title") and (h.get("url") or h.get("story_url"))
        ]
    except Exception:
        return []


def trending(query: str) -> list[dict]:
    try:
        r = requests.get(
            f"{HN_API}/search",
            params={
                "query": query,
                "tags": "story",
                "hitsPerPage": 30,
                "numericFilters": "points>10,num_comments>5",
            },
            headers=UA,
            timeout=10,
        )
        r.raise_for_status()
        hits = r.json().get("hits") or []
        return [
            _map_hit(h)
            for h in hits
            if h.get("title") and (h.get("url") or h.get("story_url"))
        ]
    except Exception:
        return []
